dqvm/src/apply-distributed-gate.lisp

```diff
@@ -0,0 +1,84 @@
+;;; apply-distributed-gate.lisp
+;;;
+;;; Author: Juan M. Bello-Rivas
+
+(in-package :dqvm2)
+
+(declaim (optimize (speed 3) (safety 0) (debug 0)))
+
+(defparameter *default-blocks-per-chunk* (expt 2 16)
+  "Number of blocks for the ranks to collectively work on during a single step within APPLY-DISTRIBUTED-GATE.")
+
+(defun apply-distributed-gate (qvm instr &key (blocks-per-chunk *default-blocks-per-chunk*))
+  (with-slots (number-of-processes number-of-qubits permutation amplitudes scratch)
+      qvm
+
+    (let* ((next-permutation (qubit-permutation instr))
+           (addresses (addresses qvm))
+           (block-size (block-size addresses))
+           (blocks-per-process (blocks-per-process addresses))
+           (requests (make-instance 'requests :total (* 2 number-of-processes))))
+
+      (loop :for block-index :from 0 :below (1+ blocks-per-process) :by blocks-per-chunk
+            :for start-offset := (* block-index block-size)
+            :for end-offset := (* (+ block-index blocks-per-chunk) block-size) :do
+
+              (non-blocking-receive qvm next-permutation start-offset end-offset requests)
+
+              (non-blocking-send qvm next-permutation start-offset end-offset requests)
+
+              ;; XXX Move call to COMPUTE-MATRIX-VECTOR-PRODUCTS here.
+              (wait-all requests))
+
+      (compute-matrix-vector-products (quil:gate-matrix instr) scratch amplitudes)
+
+      (update-permutation addresses next-permutation))))
+
+(defun make-arrays (n)
+  "Create N empty adjustable arrays."
+  (flet ((make-empty-adjustable-array ()
+           (make-array 0 :element-type '(unsigned-byte 32) :adjustable t :fill-pointer 0)))
+
+    (make-array n :initial-contents (loop :repeat n :collect (make-empty-adjustable-array)))))
+
+(defun non-blocking-receive (qvm next-permutation start-offset end-offset requests)
+  "Iterate over the addresses that should be in the current chunk after applying the next instruction. Start requests to receive the required amplitudes from the ranks where they are currently stored."
+  (let* ((addresses (addresses qvm))
+         (all-recv-offsets (make-arrays (number-of-processes addresses))))
```

Comment:
is there any bound on the size of this array? If so, can we make the arrays non-adjustable, but still have a fill pointer?